### ISSUE TITLE
Migrate a Json[] prisma 1 scalarlist to prisma 2

### ISSUE THREAD

**Issue by Aldrian** [prisma/upgrade#49]
Hi ! Related to #28, I am working in the meantime on a custom PostgreSQL script in order to migrate those relations to proper array types.

I used [the documentation](https://www.prisma.io/docs/guides/upgrade-guides/upgrade-from-prisma-1/schema-incompatibilities-postgres#scalar-lists-arrays-are-maintained-with-extra-table) in order to migrate string[] types, but I'm having an issue figuring out how to migrate a json[] type

Here's a prisma 1 subset:

```
type Live {
  id: ID! @id
  chapters: [Json] @scalarList(strategy: RELATION)
}

```

And the related prisma 2 model:

```
model Live {
  id                    String          @default(cuid()) @id
  Live_chapters         Live_chapters[]
}

model Live_chapters {
  nodeId   String
  position Int
  value    String
  Live     Live   @relation(fields: [nodeId], references: [id])
  @@id([nodeId, position])
}
```

However, the value field in `Live_chapters`, is String, not JSON, so I'm having a hard time finding a way to cast this into JSON in order to populate the new "chapters" column

Any pointers?

**Comment by matthewmueller** [comment]
And you expect it to look like this?

```prisma
model Live {
  id                    String          @default(cuid()) @id
  Live_chapters         Json[]
}
```

Also, do you have any null values in your JSON list?

**Comment by Aldrian** [comment]
I forgot to include the desired result! 
chapters Json[] would be perfect, in order to match the prisma 1 schema.

chapters[] is sometime empty, if that's what you are asking :slightly_smiling_face:

**Comment by matthewmueller** [comment]
Okay this should be fixed by #57. Please give it a try in `0.0.33`.

**Comment by Aldrian** [comment]
After talking with @pantharshit00 for the last few days on Slack, it seems that there is an issue with the generated SQL statements.

Here's a subset of my prisma 1 & 2 as a reference:

```yaml
type Media {
  id: ID! @id
  description: Json
  mediaInfos: [Json]! @scalarList(strategy: RELATION)
  chapters: [Json] @scalarList(strategy: RELATION)
}
```
```prisma
model Media {
  id                    String          @default(cuid()) @id
  description           Json?
  mediaInfos            Json[]
  chapters              Json[]
}
```

After migrating, if I try to access to a media node, like this for example

```javascript
encodingMedias: (node: any, args: any, ctx: Context) =>
    ctx.prisma.user
      .findOne({ where: { id: node.id } })
      .medias({ where: { isEncodingValidated: false } }),
```

The request fails with the following error: 

```
SyntaxError: Unexpected token , in JSON at position 475
  at JSON.parse (<anonymous>)
  at mapJson  
```

And when I look at the data using tablePlus, I can see that the json data is stringified.

If I select manually the fields, not incuding the Json[], the request succeeds.

Here is an sql statements for one of the json[] fields

```sql
ALTER TABLE "default$default"."Media" ADD COLUMN "mediaInfos" json[];
    UPDATE "default$default"."Media" u
      SET "mediaInfos" = t."value"::json[]
    FROM (
      SELECT "nodeId", array_agg(value ORDER BY position) as value
      FROM "default$default"."Media_mediaInfos"
      GROUP BY "nodeId"
    ) t
    WHERE t."nodeId" = u."id";
    DROP TABLE "default$default"."Media_mediaInfos";
```

@pantharshit00 says that you should use json_agg instead of array_agg here.

I tried re-migrating from prisma 1, and changing these statements, but the issue remains.

**Comment by matthewmueller** [comment]
@Aldrian did changing `array_agg` to `json_agg` work for you? Or are you still facing this issue?

**Comment by pantharshit00** [comment]
Hey, @Aldrian If you can paste some row from the database(not as a result from the graphql api) here which needs to combined into a json array, that will be helpful for a reproduction. So far in our conversation, I believe you pasted the json that would need to combines from API response. What we need here is a db representation.

If you can share a sql dump of your database with some rows, that would be even better as we are still guessing some details here.

**Comment by Aldrian** [comment]
@matthewmueller I somewhat ended up with no data for mediaInfos, and chapters were making it crash.

@pantharshit00 Just to be clear, you need some sql dumps right after the upgrade process done with prisma upgrade?

**Comment by pantharshit00** [comment]
@Aldrian No, I want data before the upgrade so that we can reproduce this reliably and maybe add it to our test suite.

We can also figure out a stable path for upgrade that way.

**Comment by Aldrian** [comment]
Here are two gists, one for [chapters](https://gist.github.com/Aldrian/27a8677095ce258b88cc22c9c348c59a), one for [mediaInfos](https://gist.github.com/Aldrian/7956fd7304367dbab4df7bdffdcdc32d)


Is this what you needed?

**Comment by Aldrian** [comment]
Any news on this? Was the export what you guys needed?

**Comment by pantharshit00** [comment]
Yes, @Aldrian that is the required information. I will look into this now.

**Comment by Aldrian** [comment]
Would it be possible to get an ETA on this? I think this is the last feature missing for me and I'd really need to complete the migration before the end of the next week

**Comment by pantharshit00** [comment]
Hey @Aldrian 

Sorry for the late reply. There was lot of work this week on the main Prisma repo so I didn't really had time to look here.

I managed to figure it out. 

We will need to use `array_append` and `jsonb_aggregate` in order migrate `Json[]` as the engine is storing the array data as the first element of the `jsonb[]` array. This was a bit unexpected but that is how the engine is working otherwise it can't handle the conversion.

```sql
ALTER TABLE "default$default"."Media" ADD COLUMN "mediaInfos" jsonb[];	

UPDATE
	"default$default"."Media" u
SET
	"mediaInfos" = array_append("mediaInfos", value)
FROM (
	SELECT
		"nodeId",
		jsonb_agg(value::jsonb
		ORDER BY
			"position") AS "value"
	FROM
		"default$default"."Media_mediaInfos"
	GROUP BY
		"nodeId") t
WHERE
	t. "nodeId" = u. "id"; 
```


@matthewmueller is on holiday still September 7th so this won't make into `prisma-upgrade` but I will encourage you to try this in place of queries which current `prisma-upgrade` generates. Just replace the tables.


After I used the above sql I was able to query the data:
![image](https://user-images.githubusercontent.com/22195362/92264893-94923300-eefc-11ea-8e39-08efda7b3ec7.png)

**Comment by pantharshit00** [comment]
Additional thing which came out.

We will also need to handle null rows and convert them to an empty array:
```sql
UPDATE
	"default$default"."Media"
SET
	"mediaInfos" = '{"[]"}'
WHERE
	"mediaInfos" is NULL;
```

**Comment by matthewmueller** [comment]
From Slack, @Aldrian confirmed this is working on his codebase. 

I'll add this to upgrade soon.

**Comment by Aldrian** [comment]
It is indeed working, but I had to migrate all the null data to `'{"[]"}'` in my sql statements, and be careful to set mediaInfos to `[]` on my `prisma.media.create`, since prisma client cannot handle null Json[] fields

**Comment by geosigno** [comment]
I used this solution provided:

```
mediaInfos: { set: [] },
chapters: { set: [] },
```

It worked correctly until I upgraded to Prisma 2.9.0. Now i get the below error:

`[GraphQLError: Error: Failed to validate the query: Unable to match input value to any allowed input type for the field. Parse errors: [Query parsing/validation error at Mutation.createOneMedia.data.MediaCreateInput.mediaInfos.MediaCreatemediaInfosInput.set: Value types mismatch. Have: String("[]"), want: Json, Query parsing/validation error at Mutation.createOneMedia.data.MediaCreateInput.mediaInfos Value types mismatch. Have: Object({"set": String("[]")}), want: Json] at Mutation.createOneMedia.data.MediaCreateInput.mediaInfos] {
`

Any idea how to fix this?

**Comment by pantharshit00** [comment]
@geosigno Did you also convert all your null fields to empty arrays?

**Comment by pablosat** [comment]
@geosigno since the new prisma client can handle arrays in `Json` fields, try executing this 
```sql
ALTER TABLE "default$default"."Media" ADD COLUMN "mediaInfos" jsonb;	
UPDATE
	"default$default"."Media" u
SET
	"mediaInfos" = t."value"::jsonb
FROM (
	SELECT
		"nodeId",
		jsonb_agg(value::jsonb
		ORDER BY
			"position") AS "value"
	FROM
		"default$default"."Media_mediaInfos"
	GROUP BY
		"nodeId") t
WHERE
	t. "nodeId" = u. "id"; 
```
instead of
```sql
ALTER TABLE "default$default"."Media" ADD COLUMN "mediaInfos" jsonb[];	

UPDATE
	"default$default"."Media" u
SET
	"mediaInfos" = array_append(u. "mediaInfos", value)
FROM (
	SELECT
		"nodeId",
		jsonb_agg(value::jsonb
		ORDER BY
			"position") AS "value"
	FROM
		"default$default"."Media_mediaInfos"
	GROUP BY
		"nodeId") t
WHERE
	t. "nodeId" = u. "id";
```

and this is no longer necessary:
```sql
UPDATE
	"default$default"."Media"
SET
	"mediaInfos" = '{"[]"}'
WHERE
	"mediaInfos" is NULL;
```

at least this worked for me :)

**Comment by geosigno** [comment]
Thanks for this, I am trying hard but the thing is, I already executed the first SQL on my DB :

```sql
ALTER TABLE "default$default"."Media" ADD COLUMN "mediaInfos" jsonb[];	

UPDATE
	"default$default"."Media" u
SET
	"mediaInfos" = array_append(u. "mediaInfos", value)
FROM (
	SELECT
		"nodeId",
		jsonb_agg(value::jsonb
		ORDER BY
			"position") AS "value"
	FROM
		"default$default"."Media_mediaInfos"
	GROUP BY
		"nodeId") t
WHERE
	t. "nodeId" = u. "id";

UPDATE
	"default$default"."Media"
SET
	"mediaInfos" = '{"[]"}'
WHERE
	"mediaInfos" is NULL;
```

And because of my lack of SQL skills, I don't know how to adapt your new SQL to my current DB, any help? :)
